compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java

```diff
@@ -1578,13 +1598,60 @@ private STToken readTokenInBracedContentInInterpolation() {
      */
 
     private STToken readDocumentationToken() {
-        int nextChar = peek();
-        if (nextChar == LexerTerminals.PLUS) {
+        // Look ahead and see if next non-trivial char is a plus char or a hash char.
+        // If it is a plus char, process trivial chars as leading trivia of the plus token.
+        // If it is a hash char, look ahead and see if it is followed by a deprecation literal.
+        // Else, let trivial chars be a part of the documentation description.
+        int lookAheadCount = 0;
+        int lookAheadChar = reader.peek(lookAheadCount);
+        while (lookAheadChar == LexerTerminals.SPACE || lookAheadChar == LexerTerminals.TAB) {
+            lookAheadCount++;
+            lookAheadChar = reader.peek(lookAheadCount);
+        }
+
+        if (lookAheadChar == LexerTerminals.PLUS) {
+            return processPlusToken();
+        } else if (lookAheadChar == LexerTerminals.HASH) {
+            return processDeprecationLiteralToken(lookAheadCount);
+        } else {
+            return readDocumentationInternalToken();
+        }
+    }
+
+    private STToken processPlusToken() {
+        processLeadingTrivia();
+        reader.advance();
+        switchMode(ParserMode.DOCUMENTATION_PARAMETER);
+        return getDocumentationSyntaxToken(SyntaxKind.PLUS_TOKEN);
+    }
+
+    private STToken processDeprecationLiteralToken(int lookAheadCount) {
+        // Look ahead and see if next non-trivial char belongs to a deprecation literal.
+        // There could be spaces and tabs in between.
+        lookAheadCount++;
+        int lookAheadChar = reader.peek(lookAheadCount);
+        while (lookAheadChar == LexerTerminals.SPACE || lookAheadChar == LexerTerminals.TAB) {
+            lookAheadCount++;
+            lookAheadChar = reader.peek(lookAheadCount);
+        }
+
+        // Look ahead for a "Deprecated" word match.
+        for (int i = 0; i < 10; i++) {
+            if ((char) lookAheadChar != LexerTerminals.DEPRECATED.charAt(i)) {
+                // No match. Hence return a documentation internal token.
+                return readDocumentationInternalToken();
+            }
+            lookAheadCount++;
+            lookAheadChar = reader.peek(lookAheadCount);
+        }
+
+        // There is a match. Hence return a deprecation literal.
+        processLeadingTrivia();
+        reader.mark();
+        while (!getLexeme().endsWith(LexerTerminals.DEPRECATED)) {
```

Comment:
why do we need this while loop?
is it possible to avoid `getLexeme().endsWith(LexerTerminals.DEPRECATED)` check? Coz regex-related operations on top of java Strings (such as contains(), endsWith(), startsWith(), etc ) can be really expensive..

Reply:
Fixed with 2331d55. Thanks for pointing out expensive operation. :)

Reason for the while loop:
By code line 1648, lexer knows there's a `WS # WS Deprecated` match ahead. After calling `processLeadingTrivia()` lexer doesn't know, for how many chars, the reader should be advanced.